aggregator/src/aggregator.rs

```diff
@@ -3297,24 +3296,54 @@ impl VdafOps {
                     // To ensure that concurrent aggregations don't write into a
                     // currently-nonexistent batch aggregation, we write (empty) batch
                     // aggregations for any that have not already been written to storage.
-                    let empty_batch_aggregations = empty_batch_aggregations(
+                    let batch_aggregations_iter = BatchAggregationsIterator::new(
                         &task,
                         batch_aggregation_shard_count,
                         aggregate_share_req.batch_selector().batch_identifier(),
                         &aggregation_param,
-                        &batch_aggregations,
+                        batch_aggregations,
                     );
 
-                    let (mut helper_aggregate_share, report_count, _, checksum) =
-                        compute_aggregate_share::<SEED_SIZE, B, A>(&task, &batch_aggregations)
-                            .await
-                            .map_err(|e| datastore::Error::User(e.into()))?;
+                    let aggregate_share = Arc::new(Mutex::new(AggregateShareComputer::new(&task)));
+
+                    // Rather than awaiting all the futures at once, which can cause heap growth
+                    // proportional to the number of batch aggregations to write, put them into an
+                    // buffered stream so that they get run in groups of bounded size. #3857
+                    futures::stream::iter(
+                        batch_aggregations_iter
+                            .clone()
+                            .zip(repeat_with(|| Arc::clone(&aggregate_share)))
+                            .map(Result::Ok),
+                    )
+                    .try_for_each_concurrent(
+                        max_future_concurrency,
+                        async |((ba, is_update), aggregate_share)| {
+                            aggregate_share
+                                .lock()
+                                .await
+                                .update(&ba)
+                                .map_err(|e| datastore::Error::User(e.into()))?;
```

Comment:
As in the collection job driver, we could move this to the `is_update` arm (after inlining `ba = ba.scrubbed()` below).

Moreover, we could do the entire aggregate share computation in a separate pass over the `batch_aggregations` vector, before we pass it into the new iterator. This would avoid having to introduce a mutex at all, and would be faster both due to not needing synchronization, and due to better icache locality. (I think the new `AggregateShareComputer` abstraction may still be useful, if we later split up fetching batch aggregations over multiple database queries)

Note that while the existing `compute_aggregate_share()` was an async function, it never awaited in its existing state.